### PR TITLE
[1] BoxMap overhaul

### PR DESCRIPTION
**Pull request by April-Hannah-Lena**
This PR includes many changes to BoxMap, I will edit this comment with a more detailed description tonight. But for now, some changes:

* interval artihmetic
  ```julia
  IntervalBoxMap(f, P)
  ```
* More detailed subtypes. All of the following are now callable functions
    * `BoxMap`
    * `PointDiscretizedBoxMap`
    * `MonteCarloBoxMap`
    * `GridBoxMap`
    * `AdaptiveBoxMap`
    * `CPUSampledBoxMap`
    * `GPUSampledBoxMap`
    * `IntervalBoxMap`
* simplified interface: for each of the above types, one can now call just the name within the BoxMap command. for example
  ```julia
  BoxMap(:montecarlo, f, P)
  BoxMap(:grid, f, P)
  BoxMap(:interval, f, P)
  ```
* The cpu and gpu versions have non- autogenerated aliases:
  ```julia
  BoxMap(:cpu, f, P)
  BoxMap(:gpu, f, P)
  ```

All of the examples have been changed to reflect

### CODE REVIEW

**Comment by gaioguy** [comment]
Awesome.  When describing this, we need to make sure that the user understands the type hierarchy, e.g. in form of a figure.  Also, it did not become clear to me whether I get a grid or MonteCarlo points when choosing the CPU- or GPU-accelerated BoxMaps.

**Comment by April-Hannah-Lena** [comment]
Ok, should we keep the constructor with the Symbols then? The constructors `MonteCarloBoxMap(Val(:gpu), ...)` and `GridBoxMap(Val(:gpu), ...)` are still mentioned in the docs and work independently of the Symbol-based constructors. (Same with `Val(:cpu)`)

**Comment by gaioguy** [comment]
I think the constructors with the symbols are much cleaner in terms of syntax and I would in fact strongly prefer them over the zoo of other boxmaps. Did I get that correctly that they are less efficient in some sense, though?

**Comment by April-Hannah-Lena** [comment]
As I understood it, the initial call to `BoxMap(:symbol, ...)` cannot be resolved from the type information alone. This means that BoxMap constructions cannot be completely precompiled, leading to longer time-to-first-x. I don't think it is really that much longer though, considering that other high-performance packages like ARPACK.jl also use Symbols without problems.

File: src/boxmap_cuda.jl [hunk 1]
```diff
@@ -3,6 +3,18 @@ Base.:(*)(x, ::Type{NumLiteral{T}}) where T = T(x)
 const i32, ui32 = NumLiteral{Int32}, NumLiteral{UInt32}
 
 """
+    BoxMap(:gpu, map, domain; no_of_points) -> CPUSampledBoxMap
```
Comment (gaioguy) [inline src/boxmap_cuda.jl:6]: Should that rather be `GPUSampledBoxMap'?

File: src/boxmap_cuda.jl [hunk 2]
```diff
@@ -3,6 +3,18 @@ Base.:(*)(x, ::Type{NumLiteral{T}}) where T = T(x)
...
+
+Transforms a ``map: Q → Q`` defined on points in 
+the domain ``Q ⊂ ℝᴺ`` to a `CPUSampledBoxMap` defined
```
Comment (gaioguy) [inline src/boxmap_cuda.jl:9]: Same?

File: src/no_boxmap_cuda.jl [hunk 3]
```diff
@@ -1,6 +1,18 @@
 # This file exists to provide docstrings for the Docs website
 
 """
+    BoxMap(:gpu, map, domain; no_of_points) -> CPUSampledBoxMap
```
Comment (gaioguy) [inline src/no_boxmap_cuda.jl:4]: Same?

File: src/no_boxmap_cuda.jl [hunk 4]
```diff
@@ -1,6 +1,18 @@
...
+
+Transforms a ``map: Q → Q`` defined on points in 
+the domain ``Q ⊂ ℝᴺ`` to a `CPUSampledBoxMap` defined
```
Comment (gaioguy) [inline src/no_boxmap_cuda.jl:7]: Again.

**Comment by April-Hannah-Lena** [comment]
Ah yes. Fixed.

**Pull request merged**